course/versioning.py

```diff
@@ -577,4 +598,185 @@ def update_course(pctx):
+            result.close()
+
+    return response
+
+# }}}
+
+
+@csrf_exempt
+def git_endpoint(request, course_identifier, git_path):
+    # type: (http.HttpRequest, Text, Text) -> http.HttpResponse
+
+    auth_value = request.META.get("HTTP_AUTHORIZATION")
+
+    def unauthorized_access():
+        # type: () -> http.HttpResponse
+        realm = _("Relate direct git access")
+        response = http.HttpResponse(
+                _('Authorization Required'), content_type="text/plain")
+        response['WWW-Authenticate'] = 'Basic realm="%s"' % (realm)
+        response.status_code = 401
+        return response
+
+    user = None
+    user_token = None
+    if auth_value is None:
+        return unauthorized_access()
+
+    auth_values = auth_value.split(" ")
+    if len(auth_values) != 2:
+        return unauthorized_access()
+
+    auth_method, auth_data = auth_values
+    if auth_method == "Basic":
+        from base64 import b64decode
+        auth_data = b64decode(auth_data.strip()).decode(
+                "utf-8", errors="replace")
+        auth_data_values = auth_data.split(':', 1)
+        if len(auth_data_values) != 2:
+            return unauthorized_access()
+        username, token = auth_data_values
+        try:
+            possible_user = get_user_model().objects.get(
+                    username=username)
+        except ObjectDoesNotExist:
+            return unauthorized_access()
+
+        token_values = token.split('_', 1)
+        if len(token_values) != 2:
+            return unauthorized_access()
+
+        try:
+            int(token_values[0])
+        except ValueError:
+            return unauthorized_access()
+
+        from django.utils.timezone import now
+        from django.db.models import Q
+        from django.contrib.auth.hashers import check_password
+        now_datetime = now()
+        tokens = AuthenticationToken.objects.filter(
+            user__username=username,
+            id=token_values[0],
+            participation__course__identifier=course_identifier,
+            ).filter(
+                Q(revocation_time=None)
+                | Q(revocation_time__gt=now_datetime))
+        if tokens.count() != 1:
+            return unauthorized_access()
+
+        for user_token in tokens:
+            if check_password(token_values[1], user_token.token_hash):
+                user = possible_user
+                participation = user_token.participation
+                break
+
+    if user is None:
+        return unauthorized_access()
+
+    course = participation.course
+
+    if user_token is not None and \
+            user_token.restrict_to_participation_role is not None:
+        check_permission = user_token.restrict_to_participation_role.has_permission
+        if course != user_token.restrict_to_participation_role.course:
+            return unauthorized_access()
+    else:
+        check_permission = participation.has_permission
+
+    if not (check_permission(pperm.update_content)
```

Comment:
Should that be `edit_course` or `update_content`?

Reply:
`edit_course` IMO, because that entails that they can point the git pull URL wherever they want, which entails that they can add whichever git commits they like. With `update_content`, they're limited to the pre-specified repo, which they may or may not have write access to.